### PR TITLE
[Enhancement] Implemented horizontal compaction task for lake tablet

### PR DESCRIPTION
**Pull request by sduzh**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by sduzh** [comment]
run starrocks_be_unittest

**Comment by sduzh** [comment]
run starrocks_be_unittest

File: be/src/storage/lake/rowset.cpp [hunk 1]
```diff
@@ -102,53 +77,41 @@ StatusOr<std::vector<ChunkIteratorPtr>> Rowset::get_segment_iterators(const vect
...
-        for (auto& iter : tmp_seg_iters) {
-            auto wrapper = std::make_shared<SegmentIteratorWrapper>(std::move(iter));
-            segment_iterators.emplace_back(std::move(wrapper));
-        }
+    if (segment_iterators.empty()) {
+        return vectorized::new_empty_iterator(*segment_schema, options.chunk_size);
+    } else if (segment_iterators.size() == 1) {
+        return segment_iterators[0];
+    } else if (options.sorted) {
+        return vectorized::new_heap_merge_iterator(segment_iterators);
```
Comment (wyb) [inline be/src/storage/lake/rowset.cpp:98]: The reader will not be used in vertical compaction if not return segment iterators.

File: be/src/storage/lake/horizontal_compaction_task.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    for (auto& rowset : _input_rowsets) {
+        num_rows += rowset->num_rows();
+        num_size += rowset->data_size();
+    }
+    int64_t max_input_segs = 0;
+    if (keys_type == DUP_KEYS) {
+        max_input_segs = 1;
+    } else {
+        for (auto& rowset : _input_rowsets) {
+            max_input_segs = rowset->is_overlapped() ? rowset->num_segments() : 1;
```
Comment (wyb) [inline be/src/storage/lake/horizontal_compaction_task.cpp:45]:
```suggestion
            max_input_segs += rowset->is_overlapped() ? rowset->num_segments() : 1;
```

File: be/src/storage/lake/tablet_manager.cpp [hunk 3]
```diff
@@ -396,19 +399,39 @@ Status TabletManager::publish_version(int64_t tablet_id, int64_t base_version, i
...
+            auto f = [&](const RowsetMetadataPB& r) { return r.id() == input_id; };
+            auto it = std::find_if(find_pos, metadata->rowsets().end(), f);
+            if (it == metadata->rowsets().end()) {
+                return Status::InternalError(fmt::format("input rowset {} not found", input_id));
+            }
+            find_pos = metadata->mutable_rowsets()->erase(it);
+        }
+        // Add output rowset
+        if (op_compaction.has_output_rowset() && op_compaction.output_rowset().num_rows() > 0) {
+            auto rowset = metadata->add_rowsets();
```
Comment (wyb) [inline be/src/storage/lake/tablet_manager.cpp:427]: If it is a unique key table, can not put the new rowset directly at the end.

File: be/src/storage/lake/rowset.cpp (hunk 1 above)
Comment (sduzh) [reply be/src/storage/lake/rowset.cpp:98]: I'll add a new API to get segment iterators when needed.

File: be/src/storage/lake/horizontal_compaction_task.cpp (hunk 2 above)
Comment (sduzh) [reply be/src/storage/lake/horizontal_compaction_task.cpp]: fixed

File: be/src/storage/lake/tablet_manager.cpp (hunk 3 above)
Comment (sduzh) [reply be/src/storage/lake/tablet_manager.cpp]: fixed

**Comment by sduzh** [comment]
run starrocks_be_build

**Comment by sduzh** [comment]
run starrocks_be_unittest

**Comment by sduzh** [comment]
run all

**Comment by sduzh** [comment]
run starrocks_be_unittest

File: be/src/storage/lake/horizontal_compaction_task.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    for (auto& rowset : _input_rowsets) {
+        op_compaction->add_input_rowsets(rowset->id());
+    }
+    for (auto& file : writer->files()) {
+        op_compaction->mutable_output_rowset()->add_segments(file);
+    }
+    op_compaction->mutable_output_rowset()->set_num_rows(writer->num_rows());
+    op_compaction->mutable_output_rowset()->set_data_size(writer->data_size());
+    op_compaction->mutable_output_rowset()->set_overlapped(false);
+    return _tablet->put_txn_log(std::move(txn_log));
```
Comment (wyb) [inline be/src/storage/lake/horizontal_compaction_task.cpp:116]: Need close `writer`.

File: be/src/storage/lake/tablet_manager.cpp [hunk 5]
```diff
@@ -395,24 +398,82 @@ Status TabletManager::publish_version(int64_t tablet_id, int64_t base_version, i
...
+static Status apply_compaction_log(const TxnLogPB_OpCompaction& op_compaction, TabletMetadata* metadata) {
+    // It's ok to have a compaction log without input rowset and output rowset.
+    if (op_compaction.input_rowsets().empty()) {
+        DCHECK(!op_compaction.has_output_rowset() || op_compaction.output_rowset().num_rows() == 0);
+        return Status::OK();
+    }
+    if (UNLIKELY(!op_compaction.has_output_rowset())) {
+        return Status::InternalError("compaction log contains input rowset but no output rowset");
+    }
+    if (UNLIKELY(op_compaction.output_rowset().num_rows() == 0)) {
```
Comment (wyb) [inline be/src/storage/lake/tablet_manager.cpp:420]: Num rows maybe 0 if delete is supported.

File: be/src/storage/lake/horizontal_compaction_task.cpp (hunk 4 above)
Comment (sduzh) [reply be/src/storage/lake/horizontal_compaction_task.cpp:118]: updated

File: be/src/storage/lake/tablet_manager.cpp (hunk 5 above)
Comment (sduzh) [reply be/src/storage/lake/tablet_manager.cpp]: fixed

**Review by wyb — DISMISSED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**